server/service/schedule/schedule.go

```diff
@@ -139,126 +153,221 @@ func New(
 //
 // All jobs must be added before calling Start.
 func (s *Schedule) Start() {
-	var m sync.Mutex // protects currentStart and currentWait.
-	currentStart := time.Now()
-	currentWait := 10 * time.Second
+	var intervalStartedAt time.Time // start time of the most recent run of the scheduled jobs
+	var m sync.Mutex                // protects intervalStartedAt
 
-	getWaitTimes := func() (start time.Time, wait time.Duration) {
+	getIntervalStartedAt := func() (start time.Time) {
 		m.Lock()
 		defer m.Unlock()
 
-		return currentStart, currentWait
+		return intervalStartedAt
 	}
 
-	setWaitTimes := func(start time.Time, wait time.Duration) {
+	setIntervalStartedAt := func(start time.Time) {
 		m.Lock()
 		defer m.Unlock()
 
-		currentStart = start
-		currentWait = wait
+		intervalStartedAt = start.Truncate(time.Second)
 	}
 
-	if schedInterval := s.getSchedInterval(); schedInterval < currentWait {
-		setWaitTimes(currentStart, schedInterval)
+	stats, err := s.getStats()
+	switch {
+	case errors.Is(err, sql.ErrNoRows):
+		level.Debug(s.logger).Log("msg", "start schedule, cron stats not found")
+	case err != nil:
+		level.Error(s.logger).Log("err", "start schedule", "details", err)
+		sentry.CaptureException(err)
+		ctxerr.Handle(s.ctx, err)
+	default:
+		// ok
 	}
 
-	var g sync.WaitGroup
+	if stats == nil {
+		// use empty stats so created and updated timestamps will be the zero time
+		stats = &fleet.CronStats{}
+	}
+	setIntervalStartedAt(stats.CreatedAt)
+
+	initialWait := 10 * time.Second
+	if schedInterval := s.getSchedInterval(); schedInterval < initialWait {
+		initialWait = schedInterval
+	}
+	schedTicker := time.NewTicker(initialWait)
 
-	schedTicker := time.NewTicker(currentWait)
+	var g sync.WaitGroup
 	g.Add(+1)
 	go func() {
-		defer g.Done()
+		defer func() {
+			s.releaseLock()
+			g.Done()
+		}()
 
 		for {
-			_, currWait := getWaitTimes()
-			level.Debug(s.logger).Log("msg", "waiting", "current wait time", currWait)
+			level.Debug(s.logger).Log("waiting", "wait for next tick")
 
 			select {
 			case <-s.ctx.Done():
+				schedTicker.Stop()
 				return
 
 			case <-schedTicker.C:
 				level.Debug(s.logger).Log("waiting", "done")
 
 				schedInterval := s.getSchedInterval()
-				schedTicker.Reset(schedInterval)
-
-				newStart := time.Now()
-				newWait := schedInterval
-				setWaitTimes(newStart, newWait)
 
-				if ok := s.acquireLock(); !ok {
+				stats, err := s.getStats()
+				switch {
+				case errors.Is(err, sql.ErrNoRows):
+					level.Debug(s.logger).Log("msg", "cron stats not found")
+					stats = &fleet.CronStats{}
```

Comment:
I think we should leave sql errors within the datastore. So if there are no stats, let's just return `&fleet.CronStats{}` there and in here we can only worry about scheduler logic.

Reply:
Similarly in the first time above where you get the stats.

Reply:
Sounds good. I thought about doing at the datastore instead then talked myself out so I appreciate the nudge back in that direction.